src/features/game/quiz/GameQuizManager.ts

```diff
@@ -0,0 +1,214 @@
+      await GameGlobalAPI.getInstance().showNextLine();
+      return;
+    }
+    await GameGlobalAPI.getInstance().getGameManager().getDialogueManager().hideAll();
+    let numOfCorrect = 0;
+    for (let i = 0; i < numOfQns; i++) {
+      numOfCorrect += await this.showQuizQuestion(
+        GameGlobalAPI.getInstance().getGameManager(),
+        quiz.questions[i]
+      );
+    }
+    GameGlobalAPI.getInstance().setQuizScore(quizId, numOfCorrect);
+    await GameGlobalAPI.getInstance().showNextLine();
+    await GameGlobalAPI.getInstance().getGameManager().getDialogueManager().showAll();
+  }
+
+  /**
+   * Display a prompt before a quiz starts.
+   * Player can choose to proceed and do the quiz,
+   * or to not do the quiz and exit.
+   *
+   * @param scene The Game Manager.
+   * @returns true if the player chooses to start the quiz.
+   */
+  private async showStartPrompt(scene: Phaser.Scene) {
+    const response = await promptWithChoices(scene, startPrompt.text, startPrompt.options);
+    return response === 0;
+  }
+
+  /**
+   * Display the specific quiz question.
+   *
+   * @param scene The game manager.
+   * @param question The question to be displayed.
+   */
+  public async showQuizQuestion(scene: Phaser.Scene, question: Question) {
+    const choices = question.options;
+    const quizContainer = new Phaser.GameObjects.Container(scene, 0, 0);
+
+    const quizPartitions = Math.ceil(choices.length / 5);
+    const quizHeight = choices.length;
+
+    //Create quiz box contains quiz questions
+    const quizQuestionBox = createDialogueBox(scene);
+
+    //Create text writer to display quiz questions
+    const quizQuestionWriter = createTypewriter(scene, questionTextStyle);
+    const lineWithName = question.question.replace('{name}', this.getUsername());
+    quizQuestionWriter.changeLine(lineWithName);
+
+    GameGlobalAPI.getInstance().storeDialogueLine(lineWithName, question.speaker);
+
+    //Generate UI components for quizzes
+    const header = new Phaser.GameObjects.Text(
+      scene,
+      screenSize.x - QuizConstants.textPad,
+      QuizConstants.y,
+      questionPrompt,
+      quizTextStyle
+    ).setOrigin(1.0, 0.0);
+
+    const quizHeaderBg = new Phaser.GameObjects.Rectangle(
+      scene,
+      screenSize.x,
+      QuizConstants.y - QuizConstants.textPad,
+      QuizConstants.width * quizPartitions,
+      header.getBounds().bottom * 0.5 + QuizConstants.textPad,
+      HexColor.darkBlue,
+      0.8
+    ).setOrigin(1.0, 0.0);
+
+    const quizBg = new Phaser.GameObjects.Rectangle(
+      scene,
+      screenSize.x,
+      QuizConstants.y - QuizConstants.textPad,
+      QuizConstants.width * quizPartitions,
+      quizHeaderBg.getBounds().bottom * 0.5 + (quizHeight + 0.5) * QuizConstants.yInterval,
+      HexColor.lightBlue,
+      0.2
+    ).setOrigin(1.0, 0.0);
+
+    quizContainer.add([
+      quizBg,
+      quizHeaderBg,
+      header,
+      quizQuestionBox,
+      quizQuestionWriter.container
+    ]);
```

Comment:
Given that there is quite a bit of rendering-related code here, I'm wondering if a refactor would be a possibility - any thoughts? 

(It's fine if it's non-trivial, given the complex inter-weaving between dialogue and quizzes - a bigger refactor might be necessary later on, which would be out of scope for this PR. But if possible, we should prevent it from snowballing early on)

Reply:
It seems to be possible that we can create a method for generating all the UI components and return them as an array. Any other thoughts @reginateh